test/logic/models/sort_test.dart

```diff
@@ -0,0 +1,201 @@
+        songWith(id: 3, dateModified: 0, dateAdded: 1, title: '', artist: '', album: ''),
+        songWith(id: 4, dateModified: 0, dateAdded: -1, title: '', artist: '', album: ''),
+        songWith(id: 5, dateModified: 0, dateAdded: 0, title: 'A', artist: '', album: ''),
+        songWith(id: 6, dateModified: 0, dateAdded: 0, title: 'AA', artist: '', album: ''),
+        songWith(id: 7, dateModified: 0, dateAdded: 0, title: 'a', artist: '', album: ''),
+        songWith(id: 8, dateModified: 0, dateAdded: 0, title: 'z', artist: '', album: ''),
+        songWith(id: 9, dateModified: 0, dateAdded: 0, title: '', artist: 'A', album: ''),
+        songWith(id: 10, dateModified: 0, dateAdded: 0, title: '', artist: 'AA', album: ''),
+        songWith(id: 11, dateModified: 0, dateAdded: 0, title: '', artist: 'a', album: ''),
+        songWith(id: 12, dateModified: 0, dateAdded: 0, title: '', artist: 'z', album: ''),
+        songWith(id: 13, dateModified: 0, dateAdded: 0, title: '', artist: '', album: 'A'),
+        songWith(id: 14, dateModified: 0, dateAdded: 0, title: '', artist: '', album: 'AA'),
+        songWith(id: 15, dateModified: 0, dateAdded: 0, title: '', artist: '', album: 'a'),
+        songWith(id: 16, dateModified: 0, dateAdded: 0, title: '', artist: '', album: 'z'),
+      ];
+    });
+    final Map<SortOrder, Map<SongSortFeature, List<int>>> testCases = {
+      SortOrder.ascending: {
+        SongSortFeature.dateModified: [2, 0, 3, 4, 9, 10, 11, 12, 13, 14, 15, 16, 5, 7, 6, 8, 1],
+        SongSortFeature.dateAdded: [4, 0, 1, 2, 9, 10, 11, 12, 13, 14, 15, 16, 5, 7, 6, 8, 3],
+        SongSortFeature.title: [2, 0, 3, 4, 9, 10, 11, 12, 13, 14, 15, 16, 1, 5, 7, 6, 8],
+        SongSortFeature.artist: [0, 1, 2, 3, 4, 13, 14, 15, 16, 5, 7, 6, 8, 9, 11, 10, 12],
+        SongSortFeature.album: [0, 1, 2, 3, 4, 9, 10, 11, 12, 5, 7, 6, 8, 13, 15, 14, 16],
+      },
+      SortOrder.descending: {
+        SongSortFeature.dateModified: [1, 8, 6, 5, 7, 0, 3, 4, 9, 10, 11, 12, 13, 14, 15, 16, 2],
+        SongSortFeature.dateAdded: [3, 8, 6, 5, 7, 0, 1, 2, 9, 10, 11, 12, 13, 14, 15, 16, 4],
+        SongSortFeature.title: [8, 6, 5, 7, 1, 0, 3, 4, 9, 10, 11, 12, 13, 14, 15, 16, 2],
+        SongSortFeature.artist: [12, 10, 9, 11, 8, 6, 5, 7, 0, 1, 2, 3, 4, 13, 14, 15, 16],
+        SongSortFeature.album: [16, 14, 13, 15, 8, 6, 5, 7, 0, 1, 2, 3, 4, 9, 10, 11, 12],
+      },
+    };
+    for (final entry in testCases.entries) {
+      final sortOrder = entry.key;
+      final featureTestCases = entry.value;
+      for (final entry in featureTestCases.entries) {
+        final feature = entry.key;
+        final expectedContentList = entry.value;
+        test('Sorts songs by ${feature.name} ${sortOrder.name}', () async {
+          expect(
+            songs.toList()
+              ..sort(SongSort(feature: feature, orderAscending: sortOrder == SortOrder.ascending).comparator),
+            expectedContentList.map((i) => songs[i]).toList(),
```

Comment:
Those test cases would be much more readable, if instead of IDs we would use the objects directly

You could so see the initial list vs expected list easily

Right now it's very hard to understand them

Reply:
Image one of those tests start to fail

I imagine determining what went wrong will be a real pain, but with `expect(list, expectedList)` it will show you the diff right in the error message

Reply:
I'm not sure I understand what you mean. Here we are expecting on `List<Song>`, not `List<int>`, so a failure looks like this:
```
Expected: [
            Song:Song(16),
            Song:Song(14),
            Song:Song(13),
            Song:Song(15),
            Song:Song(8),
            Song:Song(6),
            Song:Song(5),
            Song:Song(7),
            Song:Song(0),
            Song:Song(1),
            Song:Song(2),
            Song:Song(3),
            Song:Song(4),
            Song:Song(9),
            Song:Song(10),
            Song:Song(12),
            Song:Song(11)
          ]
  Actual: [
            Song:Song(16),
            Song:Song(14),
            Song:Song(13),
            Song:Song(15),
            Song:Song(8),
            Song:Song(6),
            Song:Song(5),
            Song:Song(7),
            Song:Song(0),
            Song:Song(1),
            Song:Song(2),
            Song:Song(3),
            Song:Song(4),
            Song:Song(9),
            Song:Song(10),
            Song:Song(11),
            Song:Song(12)
          ]
   Which: at location [15] is Song:<Song(11)> instead of Song:<Song(12)>
```

Maybe you are saying that testcases should be a `Map<SortOrder, Map<SongSortFeature, List<Song>>>`, but I'm not really convinced that either creating a new song with `songWith` to match one we created in the `setUp` or referencing the song created in `setUp` will be any more readable.

Reply:
About the error message - yeah, it looks good. I thought it would look different.

> Maybe you are saying that testcases should be a Map<SortOrder, Map<SongSortFeature, List<Song>>>, but I'm not really convinced that either creating a new song with songWith to match one we created in the setUp or referencing the song created in setUp will be any more readable.

Yes, that's what I meant. I would like for me to be able to humanly read those test cases and verify that the expectations have the songs actually in the order that I would expect (by seeing the song parameters)

Thinking about it more, maybe we should even compare `song.toMap`s, because now it only compares songs by IDs - it would also make the error message more detailed, and disallow any discrepancies that might happen if we implement what I suggested above

What do you think?

Reply:
I think you are right. My fear was that by creating these huge lists of song constructors it would be just overwhelming and less readable, and while I still think it's a huge wall of text, I agree that it helps to understand how we are actually sorting.

<details><summary>Example of a album test failure</summary>

```
Expected: [
            {
              'id': 10,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': -1,
              'numberOfSongs': 0
            },
            {
              'id': 13,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': -1,
              'lastYear': null,
              'numberOfSongs': 0
            },
            {
              'id': 0,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 0,
              'numberOfSongs': 0
            },
            {
              'id': 5,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': 'A',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 0,
              'numberOfSongs': 0
            },
            {
              'id': 6,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': 'AA',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 0,
              'numberOfSongs': 0
            },
            {
              'id': 7,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': 'a',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 0,
              'numberOfSongs': 0
            },
            {
              'id': 8,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': 'z',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 0,
              'numberOfSongs': 0
            },
            {
              'id': 11,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': null,
              'numberOfSongs': 0
            },
            {
              'id': 14,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': null,
              'lastYear': null,
              'numberOfSongs': 0
            },
            {
              'id': 15,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 0,
              'numberOfSongs': 1
            },
            {
              'id': 16,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 0,
              'numberOfSongs': -1
            },
            {
              'id': 9,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 1,
              'numberOfSongs': 0
            },
            {
              'id': 12,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': 1,
              'lastYear': null,
              'numberOfSongs': 0
            },
            {
              'id': 1,
              'album': 'A',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 0,
              'numberOfSongs': 0
            },
            {
              'id': 3,
              'album': 'a',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 0,
              'numberOfSongs': 0
            },
            {
              'id': 4,
              'album': 'z',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 0,
              'numberOfSongs': 0
            },
            {
              'id': 2,
              'album': 'AA',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 0,
              'numberOfSongs': 0
            }
          ]
  Actual: [
            {
              'id': 10,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': -1,
              'numberOfSongs': 0
            },
            {
              'id': 13,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': -1,
              'lastYear': null,
              'numberOfSongs': 0
            },
            {
              'id': 0,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 0,
              'numberOfSongs': 0
            },
            {
              'id': 5,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': 'A',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 0,
              'numberOfSongs': 0
            },
            {
              'id': 6,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': 'AA',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 0,
              'numberOfSongs': 0
            },
            {
              'id': 7,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': 'a',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 0,
              'numberOfSongs': 0
            },
            {
              'id': 8,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': 'z',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 0,
              'numberOfSongs': 0
            },
            {
              'id': 11,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': null,
              'numberOfSongs': 0
            },
            {
              'id': 14,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': null,
              'lastYear': null,
              'numberOfSongs': 0
            },
            {
              'id': 15,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 0,
              'numberOfSongs': 1
            },
            {
              'id': 16,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 0,
              'numberOfSongs': -1
            },
            {
              'id': 9,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 1,
              'numberOfSongs': 0
            },
            {
              'id': 12,
              'album': '',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': 1,
              'lastYear': null,
              'numberOfSongs': 0
            },
            {
              'id': 1,
              'album': 'A',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 0,
              'numberOfSongs': 0
            },
            {
              'id': 3,
              'album': 'a',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 0,
              'numberOfSongs': 0
            },
            {
              'id': 2,
              'album': 'AA',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 0,
              'numberOfSongs': 0
            },
            {
              'id': 4,
              'album': 'z',
              'albumArt': 'albumArt_albumArt_albumArt',
              'artist': '',
              'artistId': 0,
              'firstYear': 0,
              'lastYear': 0,
              'numberOfSongs': 0
            }
          ]
   Which: at location [15]['id'] is <2> instead of <4>
```
</details>

Reply:
Thanks, that is now so much better